### ISSUE TITLE
Unbounded growth of scheduler queue

### ISSUE THREAD

**Issue by sipa** [bitcoin/bitcoin#14289]
In `validation.cpp`, both `ConnectTip` and `DisconnectTip` invoke a `GetMainSignals()...` callback, which schedules a future event though the CScheduler interface.

These functions are called in 3 places:
- `ActivateBestChainTipStep`
- `InvalidateBlock`
- `RewindBlockIndex`

The first of these 3 prevents the scheduler queue from growing unboundedly, by limiting the size to 10 in `ActivateBestChainTip`. The other two however do not, and @Sjors discovered that doing a giant `invalidateblock` RPC call will in fact blow up the node's memory usage (which turns out to be caused by the queue events holding all disconnected blocks in memory).

I believe there are several improvements necessary.

1. (short term symptom fix) If this issue also appears for `RewindBlockIndex`, we must fix it before 0.17, as this may prevent nodes from upgrading pre-0.13 code. I think this is easy, as this function is called prior to normal node operation, so it can easily be reworked to release `cs_main` intermittently and drain the queue.

2. (long term fix) The use of background queues is a means to increase parallellism, not a way to break lock dependencies (as @gmaxwell pointed out privately to me, whenever a background queue is used to break a dependency, attempts to limit its size can turn into a deadlock). One way I think is to have a debug mode where the background scheduler immediately runs all scheduled actions synchronously, forcing all add-to-queue calls to happen without locks held. Once everything works in such a mode, we can safely add actual queue size limits as opposed to ad-hoc checks to prevent the queue from growing too large.

3. (orthogonal memory usage reduction) It's unfortunate that our `DisconnectBlock` events hold a shared pointer to the block being disconnected. Ideally there is a shared layer that only keeps the last few blocks in memory, and releases and reloads them from disk when needed otherwise. This wouldn't fix the unbounded growth issue above, but it would reduce the impact. It's also independently useful to reduce the memory during very large reorgs (where we can't run callbacks until the entire thing completed).

**Comment by MarcoFalke** [comment]
See also  "Unbounded reorg memory usage" #9027  about resource usage of blocks kept in memory and the wallet lagging behind too much.

**Comment by Sjors** [comment]
Even if it's not a regression, the deeper SegWit activation is burried, the more strain is put on `RewindBlockIndex`. Perhaps at some depth the node should not automatically upgrade, but promt the user whether they want to do that, or if they prefer to delete and resync (probably faster), or even ignore SegWit until some later moment.

**Comment by gmaxwell** [comment]
It's a regression. At least before a walletless, txindexless node could happily invalidate down to 0 without causing extreme memory usage. It doesn't now.  The memory usage never goes down, even after returning back to the tip.   64GB of usage is hit after invalidating only 40k blocks too.

**Comment by Sjors** [comment]
I'm syncing a v0.13.0 node now on AWS (i3.2xlarge) and plan to upgrade to upgrade to v0.17.0rc4 once it's near or at the tip, to see how memory behaves.

**Comment by Sjors** [comment]
After syncing v0.13 I upgraded to v0.17.0rc. It first upgraded the UTXO database as expected and then started rolling back blocks as expected.

Cache 1.5 GB, RAM 12.9 GB:
![schermafbeelding 2018-09-25 om 21 36 18](https://user-images.githubusercontent.com/10217/46044659-b1703a00-c11b-11e8-9d65-4dde6ceba6aa.png)

Cache 2.4 GB, RAM 20 GB:
<img width="1440" alt="schermafbeelding 2018-09-25 om 21 44 31" src="https://user-images.githubusercontent.com/10217/46044660-b1703a00-c11b-11e8-9109-6ff8f2381ad4.png">

No graceful stopping this either:
<img width="460" alt="schermafbeelding 2018-09-25 om 21 44 57" src="https://user-images.githubusercontent.com/10217/46044662-b1703a00-c11b-11e8-8ff7-76864e7bfdd1.png">

**Comment by Sjors** [comment]
Using a backup of the fully synced v0.13.0 database, I tried v0.16.3. It also first upgraded the UTXO database as expected and then started rolling back blocks as expected.

Cache  1.4 GB, RAM 11.2  GB at height 533K:
<img width="2131" alt="533k" src="https://user-images.githubusercontent.com/10217/46087458-ab756a00-c1aa-11e8-9226-2f7a4a79580b.png">

**Comment by Sjors** [comment]
v0.15.2 does not have this bug. It rolled back to 533K with only 3.2 GB of RAM usage (it loaded the full UTXO database during the database upgrade). It's still pretty slow, so I don't think automatic rollback is a great user experience.

**Comment by MarcoFalke** [comment]
> Using a backup of the fully synced v0.13.0 database, I tried v0.16.3. It also first upgraded the UTXO database as expected and then started rolling back blocks as expected.
> Cache 1.4 GB, RAM 11.2 GB at height 533K:

Is that with wallet and txindex disabled?

**Comment by Sjors** [comment]
No wallet, no indexes.

**Comment by MarcoFalke** [comment]
Which weakly indicates that this is not a regression in 0.17.0

**Comment by Sjors** [comment]
@marcofalke it's a regression in 0.16, based on the above tests.

**Comment by MarcoFalke** [comment]
Moved to 0.17.1, as this is a known issue: https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.17.0-Release-notes/_history

**Comment by laanwj** [comment]
It's not realistic to solve this entirely before the 0.18 release, and also not required as this is not a regression for 0.18—so I'm moving the milestone. Any fixes in this regard should of course be backported to the 0.18 branch.

**Comment by MarcoFalke** [comment]
Could someone update the current status of this issue? I think it is hard to follow, since some if it is already solved, but other parts weren't yet addressed.

**Comment by Sjors** [comment]
The specific instance I found with `invalidateblock`, and rewinding the chain in general, is solved.

**Comment by laanwj** [comment]
In how far is this a blocker for 0.19? Should we push forward the milestone again?

**Comment by MarcoFalke** [comment]
What exactly is left to do here?

**Comment by MarcoFalke** [comment]
What exactly is left to be done here?

**Comment by promag** [comment]
Ping.
